### PR TITLE
Added connectRetryCount and connectRetryInterval to SQLServerDatasource

### PR DESCRIPTION
**Pull request by lilgreenbird**

### CODE REVIEW

**Review by v-zhangw — APPROVED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 1]
```diff
@@ -1066,4 +1066,24 @@
      *        String value for 'maxResultBuffer'
      */
     void setMaxResultBuffer(String maxResultBuffer);
+
+    /**
+     * Sets the number of reconnection attempts if there is a connection failure
```
Comment (karinazhou) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1071]: Please add the param to the comment.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 2]
```diff
@@ -1066,4 +1066,24 @@
...
+     */
+    void setConnectRetryCount(int count);
+
+    /**
+     * Returns the number of reconnection attempts if there is a connection failure
+     */
+    int getConnectRetryCount();
+
+    /**
+     * Sets the number of seconds between each connection retry attempt
```
Comment (karinazhou) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1081]: Please add the param to the comment.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 3]
```diff
@@ -1066,4 +1066,24 @@
...
      */
     void setMaxResultBuffer(String maxResultBuffer);
+
+    /**
+     * Sets the number of reconnection attempts if there is a connection failure
+     */
+    void setConnectRetryCount(int count);
+
+    /**
+     * Returns the number of reconnection attempts if there is a connection failure
```
Comment (karinazhou) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1076]: NIT:  this is the single-line comment so I think `//` should be used here.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 4]
```diff
@@ -1066,4 +1066,24 @@
...
+     */
+    int getConnectRetryCount();
+
+    /**
+     * Sets the number of seconds between each connection retry attempt
+     */
+    void setConnectRetryInterval(int interval);
+
+    /**
+     * Returns the number of seconds between each connection retry attempt
```
Comment (karinazhou) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1086]: NIT:  this is the single-line comment so I think `//` should be used here.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java (hunk 3 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1076]: javadocs

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java (hunk 4 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1086]: javadocs

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/TimeoutTest.java [hunk 5]
```diff
@@ -154,6 +155,33 @@ public void testConnectRetryServerError() {
...
     }
 
+    // Test connect retry for database error using Datasource
+    @Test
+    public void testConnectRetryServerErrorDS() {
+        long timerEnd = 0;
+        long timerStart = System.currentTimeMillis();
+
+        // non existent database with interval < loginTimeout this will generate a 4060 transient error and retry
+        int connectRetryCount = new Random().nextInt(255);
```
Comment (karinazhou) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/TimeoutTest.java:165]: Random().nextInt(255) excludes 255 so this value won't be tested. I think we need to use Random().nextInt(256) here.

**Review by peterbae — APPROVED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/MaxResultBufferCounter.java [hunk 6]
```diff
@@ -20,6 +20,12 @@
     private long counter = 0;
     private final long maxResultBuffer;
 
+    /**
+     * Sets the value of the max result buffer counter
```
Comment (karinazhou) [inline src/main/java/com/microsoft/sqlserver/jdbc/MaxResultBufferCounter.java:24]: I see we use `Constructs` for constructors in the comments.

**Review by karinazhou — APPROVED** [review]

**Review by karinazhou — APPROVED** [review]

**Review by peterbae — APPROVED** [review]

**Review by v-zhangw — APPROVED** [review]

**Pull request merged**